### ISSUE TITLE
Add new VGUI browser panel using ISteamHTMLSurface API

### ISSUE THREAD

**Issue by samuelmaddock** [Facepunch/garrysmod-requests#791]
Instead of replacing Awesomium, add an additional VGUI panel using the [ISteamHTMLSurface API](https://partner.steamgames.com/documentation/html_surface). Some benefits include:
- Evergreen browser; Steam will continue updating their underlying version of CEF
- Alleviates outdated browser issues in addons using streaming services
- Includes codecs for MP4 playback as well as DRM support (no licensing required)
- Binaries already included with Steam for all platforms
- No need to update existing game features (menu, third-party libraries, etc.)
- No need to match existing DHTML API since this would be a new VGUI panel

Hopefully this could be a viable alternative to replacing Awesomium as previously rejected in #531. I haven't followed the issues regarding OS X 32-bit support with CEF, so I'm unsure if that would also be an issue here.

**Comment by samuelmaddock** [comment]
I recently found that Source SDK 2013 contains an [updated HTML VGUI panel](https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/vgui2/vgui_controls/HTML.cpp) which uses the ISteamHTMLSurface API. Is it be possible to integrate these changes into GMod? @willox @robotboy655

**Comment by jackson-c** [comment]
Sam I Love You :smiley:

**Comment by Digiquo** [comment]
I'm here to help stop my favorite Gmod mode from dying (Cinema) and to support the longevity of Gmod in the long run. #replaceawesomium

**Comment by DemonAura** [comment]
+1 The longer this goes on, the worse it gets. Awesomium is incredibly outdated and needs to be put to rest in favor of a more capable framework.

**Comment by Sophie-bear** [comment]
Great idea, Sam. Hopefully this won't be shot down because it's very important to a lot of people and could bring about a lot of changes for the better.

**Comment by gonzalologorg** [comment]
I never played cinema and possibly i will never play, but this guy...Deserves this...I don't want something that cool to die

**Comment by jasmin12e** [comment]
+1 Please don't let things get worse I really enjoy cinema servers but everything needs an overhaul!

**Comment by SunsetsAndSushi** [comment]
+1 Don't let Cinema die.

**Comment by Daniel-Fewster** [comment]
This issue has gone on way too long, and this looks to be a sound alternative if nothing else in the efforts to leave Awesomium's curse...

**Comment by WinterPhoenix96** [comment]
Once again, what we are suggesting is **not** that this VGUI Panel replace Awesomium, especially since Garry seems very against that idea, but **an alternative, implemented separately to Awesomium**, that GMod developers could work with. It would be completely optional.

**Comment by InvisbleMario** [comment]
+1 But if someone actually manages to implement this as a pull request their would probably be a higher chance of it being accepted.

**Comment by jackson-c** [comment]
@garrynewman

**Comment by InvisbleMario** [comment]
@willox @robotboy655

**Comment by willox** [comment]
We're not against the idea, but don't expect it for the next update

**Comment by WinterPhoenix96** [comment]
I'm genuinely glad to hear that, at the very least, you're not against the idea. If you can, please keep us posted. Thanks!

**Comment by El13B** [comment]
I'm here to support Winter ^^ good luck Winter!

**Comment by jbuug** [comment]
So what you mean is update the interface on the game, if so im rooting for ya

**Comment by Kamshak** [comment]
Any news on this? Is there like a roadmap or something? Would be amazing to see this implemented, also as it is the most "Thumbed up" issue in here.

**Comment by mrpotatofactory** [comment]
You guys have all my support.

**Comment by uRandomAlex** [comment]
6 months since this request was born. [sarcasm]Unfortunately not 1 year. Hope it will stay unimplemented for at least 6 months more, so I can say "Happy Birthday!".[/sarcasm]

**Comment by WinterPhoenix96** [comment]
Two video services have now broken in the last month that were previously supported in Cinema/Media Player. They were Vimeo and Twitch, and both are no longer supported due to Awesomium's lack of newer JavaScript and HTML5 functionality.

Furthermore, Google Chrome is now (as of version 55, rolling out as of December 2nd) officially shipping [Flash Player disabled by default](https://9to5google.com/2016/12/01/chrome-55-stable-adobe-flash-html5/), which will only expedite the process of the web as a whole abandoning Adobe Flash Player in favor of HTML5.

To be frank, Cinema and Media Player are looking to flat out lose support for all their video services very soon if this issue is not addressed. At the time of writing, this is the most Thumbed Up Issue on this GitHub Repo, with second place, #745, having _40 less_, so why has it not been treated as a priority up until this point?

So **GMod Devs**, once again, for the love of god please implement this.

Thank you.

**Comment by uRandomAlex** [comment]
@WinterPhoenix96 Don't worry, they will try to do something when Cinema dies completely.

**Comment by jackson-c** [comment]
Let's do a binary add-on, lol.

**Comment by DoomandGloom** [comment]
I just hope the gmod devs see how passionate we are about the cinema game mode and how much creativity and effort that goes into each server and what effect it would have on our communities when/if it dies out.

Thank you for raising these discussions and I hope a future solution comes.

**Comment by Noiwex** [comment]
yes do this

**Comment by Noiwex** [comment]
Flash is getting slowly phased out. More and more websites turn their sides to HTML5. If @willox @robotboy655 @garrynewman won't do anything, it will be a catastrophe for gmod.

**Comment by Noiwex** [comment]
What happens when awesomium in gmod meets some critical exploit? Will gmod team monkey patch the issue, or update awesomium, or do anything at all?

**Comment by Noiwex** [comment]
I cannot deny that there IS some exploit used in the wild that could compromise your system by placing some DLLs in your bin or steam folder.

**Comment by samuelmaddock** [comment]
I think it's worth reiterating that the intention of this request is to add a new browser VGUI control rather than replacing Awesomium. The ISteamHTMLSurface API isn't 100% compatible with the functionality Awesomium provides. To name a few incompatibilies: 1. no callbacks for console logged messages 2. no support for binding to JavaScript functions. These features would require work on behalf of Valve.

Getting rid of Awesomium would be nice—ISteamHTMLSurface API might not fit the requirements to replace it though.

**Comment by Noiwex** [comment]
Or like, it would be great if someone made @garrynewman sober and he approved works on CEF.

**Comment by mrpotatofactory** [comment]
CEF does not work on Mac. Then again, why not force Mac users to stay with Awesomium and the rest of us use CEF?

**Comment by Bo98** [comment]
Older OS versions are also a problem.

**Comment by Noiwex** [comment]
GMod works TOTALLY horrible on platforms that aren't windows. It's impossible to play on Mac or Linux without crashes every 10 minutes.

**Comment by Noiwex** [comment]
Does Rust still support DX9 much? Hell no, they deprecated the shit out of it. Fuck older OSes. Fuck Valve's ToGL wrapper.

**Comment by Bo98** [comment]
Rust still runs in DX9 fine afaik.

**Comment by uRandomAlex** [comment]
@Noiwex I remember playing vanilla TTT totally fine on Linux for 3 hours without any crashes.

**Comment by Noiwex** [comment]
Try something more heavy than TTT

**Comment by Zenshu** [comment]
sooo... news?

**Comment by Noiwex** [comment]
no available facepunch stuff it seems

**Comment by Adam759** [comment]
This is so ridiculous that something so crucial with Garrysmod is just being ignored and pushed aside. I wish they cared more to actually listen.

**Comment by willox** [comment]
This can't happen because browsers created with the ISteamHTMLSurface seem to share their session with the Steam Overlay's browser.

**Comment by jackson-c** [comment]
So pages created on steamcommunity.com have the user's login and are vulnerable to extreme XSS? Understandable, but we still need a new embedded browser one way or another.

**Comment by Cpt-Derp** [comment]
@Noiwex I might be wrong, but the crashing every 10 minutes may have something to do with OpenGL. It has limitations with 32-bit, with how it handles texture memory (3.5 GB limit). 

I've noticed that setting the game to lower graphics settings significantly reduces the incidence of crashes and observing VRAM usage, Garry's Mod appears to crash when the 3.5 GB mark is reached.

Probably no way to fix that unless Garry's Mod switches to 64 bit, which might not be feasible.

**Comment by Noiwex** [comment]
Confirmed. http://i.imgur.com/3cH7wTX.png
I've set motd url to https://steamcommunity.com/my

**Comment by jackson-c** [comment]
I mean... you could do something like blacklisting RunJavascript on steam* domains.

**Comment by ZokeysEPN** [comment]
This is getting annoying, please fix it!

**Comment by WinterPhoenix** [comment]
Please see this issue for more up to date discussion: https://github.com/Facepunch/garrysmod-requests/issues/531

**Comment by unknown-gd** [comment]
@robotboy655 Please if you still care about the game please update the damn web browser already, we've been unable to properly use media content in the game for an infinitely long 10 years now. It's literally killing the game, the guys at @solsticegamestudios have done an incredible work and YOU JUST IGNORE IT. Just accept this [PR](https://github.com/Facepunch/gmod-html/pull/3) already and I guarantee you this will be a new second holiday for all players.

**Comment by Cloudwalk9** [comment]
@unknown-gd Are you saying they're still using Awesomium on the main branch? Or they're not updating the CEF binaries that even fucking Gentoo builds optional binaries for?

Or both?

**Comment by Cloudwalk9** [comment]
@DarthTealc are you fucking kidding me? I was in 8th and 9th grade in 2012. The Flash YouTube player was still preferred on slower machines where HTML5 was still relatively nascent.

**Comment by Cloudwalk9** [comment]
Like that's actually irresponsible. You buy the game, new player, unaware one of its major components has accumulated CVEs over a period of 12 years and it's not immediately obvious a much more secure version is under the betas but is still not updated properly. Even worse that the game still receives patches lmao

**Comment by Cloudwalk9** [comment]
Like that's actually irresponsible. ***[There's over 1500 CVEs since Chromium 18, which Awesomium is based on](https://cve.mitre.org/cgi-bin/cvekey.cgi?keyword=chromium)***

You buy the game, new player, unaware one of its major components has accumulated that many vulnerabilities over a period of 12 years and it's not immediately obvious a much more secure version is under the betas but is still not updated properly and has hundreds of them still. Even worse that the game still receives patches lmao

**Comment by unknown-gd** [comment]
> [@unknown-gd](https://github.com/unknown-gd) Are you saying they're still using Awesomium on the main branch? Or they're not updating the CEF binaries that even fucking Gentoo builds optional binaries for?
> 
> Or both?

Main game branch. Players often don't want to switch to beta, another problem is that all cinema servers and modes require H.264/265 codec and force players to modify their clients via CodecFix to share video, which I also think is not normal and not safe. I don't think updating this old crap is complicated enough to put it off for 10 years.

**Comment by unknown-gd** [comment]
> [@unknown-gd](https://github.com/unknown-gd) Are you saying they're still using Awesomium on the main branch? Or they're not updating the CEF binaries that even fucking Gentoo builds optional binaries for?
> 
> Or both?

Both, but the main game branch is just dead, you can't run anything on it anymore, even google will refuse to load the page normally. Also players often don't want to upgrade to the beta and another problem is that all the `Cinema` servers and gamemode require `H.264`/ `H.265` codecs and force players to modify their clients via CodecFix to watch video together, which in my opinion is not normal or safe. And I don't think updating this old crap is hard enough to put it off for 10 freaking years, they could have hired a third-party developer to do this a long time ago.

**Comment by swampservers** [comment]
What is apparent from Facepunch's continued lack of action on this issue is that they simply don't care about maintaining GMod. All of their effort is going towards S&Box.

While S&Box looks cool on the surface, as a developer who is highly invested in GMod who has been repeatedly frustrated by Facepunch's utter negligence of this 10-year issue that majorly affects both usability and security, I would find it very hard to justify investing significant time into their new platform. If Garry cannot even be bothered to throw a few thousand dollars at this problem to get it resolved, then he has clearly shown that he just doesn't care about people who build off his work.

**Comment by garryspins** [comment]
Just do _something_ Rubat, throw us a bone, you know its an issue, stop plugging your ears, atleast give us a reason.

**Comment by swampservers** [comment]
This is an issue that affects more than just Cinema & media players. About a year ago I spent a bunch of time upgrading my website, which is used for the in-game MOTD, to bootstrap 5.0 (from bootstrap 3.0) and had to undo it all because it just doesn't work on Awesomium.

At this point I don't think there is much purpose to continuing to scream about it in GMod related channels because they clearly aren't motivated to fix it. They might be motivated to fix it, however, if they realize that this *abandonment* reflects poorly on Facepunch as a whole, which makes their new project S&Box less appealing as I described.
